File: PrefixSum/3739. Count Subarrays With Majority Element II/solution.py
from collections import Counter
from itertools import accumulate
from typing import List

class Solution:
    def countMajoritySubarrays(self, nums: List[int], target: int) -> int:
        # Map to +1 if equals target, else -1
        arr = [1 if x == target else -1 for x in nums]
        
        # Prefix sums, with pref[0] = 0
        presum = list(accumulate(arr, initial=0))

        # Coordinate compression
        vals = sorted(set(presum))
        idx = {v: i for i, v in enumerate(vals, start=1)}  # 1-based

        # Fenwick Tree
        size = len(vals)
        bit = [0] * (size + 2)

        def update(i: int, delta: int) -> None:
            while i <= size:
                bit[i] += delta
                i += i & -i

        def query(i: int) -> int:
            res = 0
            while i > 0:
                res += bit[i]
                i -= i & -i
            return res

        # Count pairs with pref[j] > pref[i]
        res = 0
        for cnt in presum:
            pos = idx[cnt]
            res += query(pos - 1)
            update(pos, 1)

        return res
    
class Solution:
    def countMajoritySubarrays(self, nums: List[int], target: int) -> int:
        """
        A subarray has target as majority
        if its balance increase is positive,
        meaning target appears more than half.

        count[x] stores how many prefixes have balance x.
        acc[x] is the prefix sum of count.

        Each step adds acc[pre - 1] to
        count all earlier prefixes with smaller balance
        these correspond to valid subarrays where target dominates.

        acc[i] means the number of times the prefix sum <= i has occurred.
        There is an offset of n+1 used in the algorithm to map index with prefix sum, which works fine even if the prefix sum goes negative.
        """
        n = len(nums)
        arr = [1 if x == target else -1 for x in nums]

        count = Counter()
        count[0] = 1

        acc = [1] + [0] * (n + n + 2)
        res = presum = 0
        for num in arr:
            presum += num
            count[presum] += 1
            acc[presum] = acc[presum - 1] + count[presum]
            res += acc[presum - 1]
        return res

File: PrefixSum/3739. Count Subarrays With Majority Element II/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    long long countMajoritySubarrays(vector<int>& nums, int target) {
        int n = nums.size();
        
        // 前綴和範圍在 [-n, n] 之間
        // 加上 offset = n + 1，將其映射到 [1, 2*n + 1] 區間，防止負數索引
        int offset = n + 1;
        vector<int> bit(2 * n + 2, 0);

        // 樹狀陣列標準操作：單點增加
        auto update = [&](int idx, int val) {
            for (; idx < bit.size(); idx += idx & -idx) {
                bit[idx] += val;
            }
        };

        // 樹狀陣列標準操作：前綴和查詢
        auto query = [&](int idx) {
            int sum = 0;
            for (; idx > 0; idx -= idx & -idx) {
                sum += bit[idx];
            }
            return sum;
        };

        long long res = 0;
        int presum = 0;

        // 初始狀態：當還沒包含任何元素時，l-1 = -1 的前綴和為 0
        // 將 0 映射後放入 BIT 中
        update(0 + offset, 1);

        for (int num : nums) {
            // 轉換為 +1 / -1
            presum += (num == target ? 1 : -1);
            
            // 我們需要找符合 previous_presum < presum 的個數
            // 即為查詢小於等於 presum - 1 的所有歷史紀錄總數
            res += query(presum - 1 + offset);
            
            // 將當前的 presum 記錄進 BIT 中
            update(presum + offset, 1);
        }

        return res;
    }
};